File: R/prepare_availability.R
library(readxl)
library(jsonlite)
library(stringr)
library(stringi)
library(purrr)
library(dplyr)

slugify_hut_id <- function(raw_hut) {
  x <- trimws(as.character(raw_hut))

  replacements <- c(
    "ß" = "ss", "ẞ" = "ss",
    "ä" = "a", "ö" = "o", "ü" = "u",
    "Ä" = "a", "Ö" = "o", "Ü" = "u"
  )

  x <- stringr::str_replace_all(x, replacements)
  x <- stringi::stri_trans_general(x, "Latin-ASCII")
  x <- tolower(x)
  x <- stringr::str_replace_all(x, "[^a-z0-9]+", "-")
  x <- stringr::str_remove(x, "^-+")
  x <- stringr::str_remove(x, "-+$")
  x
}

split_hut_country <- function(raw_hut) {
  raw_hut <- trimws(as.character(raw_hut))
  m <- stringr::str_match(raw_hut, "^(.*?),\\s*([A-Z]{2})$")

  if (!is.na(m[1, 1])) {
    list(name = m[1, 2], country = m[1, 3])
  } else {
    list(name = raw_hut, country = NA_character_)
  }
}

parse_calendar_date <- function(month_header, day) {
  if (is.na(month_header) || month_header == "" || is.na(day) || day == "") {
    return(NA_character_)
  }

  mh <- trimws(as.character(month_header))
  m <- stringr::str_match(mh, "^(\\d{1,2})/(\\d{4})$")

  if (is.na(m[1, 1])) {
    stop("Unexpected month_header: ", month_header)
  }

  month <- as.integer(m[1, 2])
  year <- as.integer(m[1, 3])
  day_val <- as.integer(day)

  res_date <- tryCatch(
    as.Date(sprintf("%04d-%02d-%02d", year, month, day_val)),
    error = function(e) NA
  )

  if (is.na(res_date)) return(NA_character_)
  format(res_date, "%Y-%m-%d")
}

normalize_status <- function(raw_status, free_places, disabled) {
  raw_status <- tolower(trimws(as.character(raw_status)))
  if (is.na(raw_status)) raw_status <- ""

  free_clean <- suppressWarnings(as.integer(free_places))
  if (length(free_clean) == 0 || is.na(free_places) || free_places == "") {
    free_clean <- NA_integer_
  }

  if (raw_status == "error") {
    return(list(status = "error", level = "error", free = free_clean))
  }

  is_disabled <- !is.na(disabled) && (
    isTRUE(disabled) || tolower(as.character(disabled)) == "true"
  )

  if (is_disabled) {
    return(list(status = "closed", level = "closed", free = free_clean))
  }

  if (raw_status == "unknown" || is.na(free_clean)) {
    return(list(status = "unknown", level = "unknown", free = NA_integer_))
  }

  if (free_clean <= 0) {
    return(list(status = "full", level = "full", free = 0L))
  }

  if (free_clean <= 3) {
    return(list(status = "available", level = "low", free = free_clean))
  }

  if (free_clean <= 9) {
    return(list(status = "available", level = "medium", free = free_clean))
  }

  list(status = "available", level = "high", free = free_clean)
}

null_if_na <- function(x) {
  if (length(x) == 0 || is.na(x)) NULL else x
}

build_availability_outputs <- function(df_rows) {
  generated_at <- format(Sys.time(), "%Y-%m-%dT%H:%M:%S")

  long_rows <- list()
  errors_list <- list()
  huts_list <- list()

  if (nrow(df_rows) == 0) {
    return(list(
      long_rows = tibble::tibble(),
      json = list(
        generated_at = generated_at,
        date_from = NULL,
        date_to = NULL,
        errors = list(),
        huts = list()
      )
    ))
  }

  for (i in seq_len(nrow(df_rows))) {
    r <- df_rows[i, ]

    raw_hut <- r$Hut
    if (is.na(raw_hut) || trimws(as.character(raw_hut)) == "") next

    raw_hut <- trimws(as.character(raw_hut))
    hut_id <- slugify_hut_id(raw_hut)

    hut_info <- split_hut_country(raw_hut)
    hut_name <- hut_info$name
    country <- hut_info$country

    raw_status <- if ("status" %in% names(r)) r$status else NA_character_
    raw_status_chr <- tolower(trimws(as.character(raw_status)))
    if (is.na(raw_status_chr)) raw_status_chr <- ""

    if (raw_status_chr == "error") {
      err_msg <- if ("error_message" %in% names(r)) r$error_message else NA_character_
      if (!hut_id %in% names(errors_list)) errors_list[[hut_id]] <- list()

      errors_list[[hut_id]][[length(errors_list[[hut_id]]) + 1]] <- list(
        hut = raw_hut,
        message = null_if_na(err_msg)
      )
      next
    }

    month_header <- if ("month_header" %in% names(r)) r$month_header else NA_character_
    day <- if ("day" %in% names(r)) r$day else NA_integer_
    iso_date <- parse_calendar_date(month_header, day)

    if (is.na(iso_date)) next

    free_places <- if ("free_places" %in% names(r)) r$free_places else NA_integer_
    disabled <- if ("disabled" %in% names(r)) r$disabled else NA
    norm <- normalize_status(raw_status, free_places, disabled)

    if (!hut_id %in% names(huts_list)) {
      huts_list[[hut_id]] <- list(
        hut = raw_hut,
        name = hut_name,
        country = null_if_na(country),
        calendar = list()
      )
    }

    day_obj <- list(
      free = if (is.na(norm$free)) NULL else norm$free,
      status = norm$status,
      level = norm$level
    )

    if ("raw_text" %in% names(r) && !is.na(r$raw_text)) day_obj$raw_text <- as.character(r$raw_text)
    if ("aria" %in% names(r) && !is.na(r$aria)) day_obj$aria <- as.character(r$aria)

    huts_list[[hut_id]]$calendar[[iso_date]] <- day_obj

    long_rows[[length(long_rows) + 1]] <- tibble::tibble(
      hut_id = hut_id,
      hut_name = hut_name,
      country = if (is.na(country)) "" else country,
      date = iso_date,
      free_places = if (is.na(norm$free)) NA_integer_ else norm$free,
      status = norm$status,
      level = norm$level,
      raw_hut = raw_hut,
      raw_status = if (is.na(raw_status)) "" else as.character(raw_status),
      raw_text = if ("raw_text" %in% names(r) && !is.na(r$raw_text)) as.character(r$raw_text) else "",
      aria = if ("aria" %in% names(r) && !is.na(r$aria)) as.character(r$aria) else "",
      disabled = if ("disabled" %in% names(r) && !is.na(r$disabled)) as.character(r$disabled) else "",
      color = if ("color" %in% names(r) && !is.na(r$color)) as.character(r$color) else ""
    )
  }

  if (length(long_rows) > 0) {
    df_long <- dplyr::bind_rows(long_rows)
    all_dates <- sort(unique(df_long$date))
  } else {
    df_long <- tibble::tibble(
      hut_id = character(), hut_name = character(), country = character(), date = character(),
      free_places = integer(), status = character(), level = character(), raw_hut = character(),
      raw_status = character(), raw_text = character(), aria = character(), disabled = character(),
      color = character()
    )
    all_dates <- character()
  }

  for (hut_id in names(huts_list)) {
    cal <- huts_list[[hut_id]]$calendar

    available_days <- purrr::keep(names(cal), function(d) {
      identical(cal[[d]]$status, "available") &&
        !is.null(cal[[d]]$free) &&
        !is.na(cal[[d]]$free) &&
        cal[[d]]$free > 0
    })
    available_days <- sort(available_days)

    days_full <- sum(purrr::map_lgl(cal, ~ identical(.x$status, "full")))
    days_unknown <- sum(purrr::map_lgl(cal, ~ identical(.x$status, "unknown")))

    if (length(available_days) > 0) {
      next_avail_date <- available_days[1]
      next_avail_free <- cal[[next_avail_date]]$free
      all_free_counts <- purrr::map_int(available_days, ~ cal[[.x]]$free)
      max_free <- max(all_free_counts)
      total_free <- sum(all_free_counts)
    } else {
      next_avail_date <- NULL
      next_avail_free <- NULL
      max_free <- NULL
      total_free <- 0L
    }

    huts_list[[hut_id]]$summary <- list(
      days_total = length(cal),
      days_available = length(available_days),
      days_full = days_full,
      days_unknown = days_unknown,
      next_available_date = next_avail_date,
      next_available_free = next_avail_free,
      max_free_places = max_free,
      total_free_place_days = total_free
    )
  }

  if (length(huts_list) > 0) {
    huts_list <- huts_list[order(names(huts_list))]
  }

  availability_json <- list(
    generated_at = generated_at,
    date_from = if (length(all_dates) > 0) all_dates[1] else NULL,
    date_to = if (length(all_dates) > 0) all_dates[length(all_dates)] else NULL,
    errors = errors_list,
    huts = huts_list
  )

  list(long_rows = df_long, json = availability_json)
}

prepare_availability <- function(
  in_xlsx = file.path("data", "calendar_results.xlsx"),
  out_json = file.path("docs", "availability.json"),
  out_csv = file.path("data", "availability_long.csv")
) {
  if (!file.exists(in_xlsx)) {
    stop("Vstupní soubor neexistuje: ", in_xlsx)
  }

  dir.create(dirname(out_json), recursive = TRUE, showWarnings = FALSE)
  dir.create(dirname(out_csv), recursive = TRUE, showWarnings = FALSE)

  df_raw <- readxl::read_xlsx(in_xlsx, sheet = 1)
  outputs <- build_availability_outputs(df_raw)

  readr::write_csv(outputs$long_rows, out_csv, na = "")

  writeLines(
    jsonlite::toJSON(outputs$json, auto_unbox = TRUE, pretty = TRUE, null = "null", na = "null"),
    out_json,
    useBytes = TRUE
  )

  num_errors <- sum(purrr::map_int(outputs$json$errors, length))

  cat("Input rows:", nrow(df_raw), "\n")
  cat("Calendar rows written:", nrow(outputs$long_rows), "\n")
  cat("Huts:", length(outputs$json$huts), "\n")
  cat("Date range:", outputs$json$date_from, "–", outputs$json$date_to, "\n")
  cat("Errors:", num_errors, "\n")
  cat("Wrote:", out_json, "\n")
  cat("Wrote:", out_csv, "\n")

  invisible(outputs)
}

if (sys.nframe() == 0) {
  args <- commandArgs(trailingOnly = TRUE)
  in_xlsx <- if (length(args) >= 1) args[1] else file.path("data", "calendar_results.xlsx")
  out_json <- if (length(args) >= 2) args[2] else file.path("docs", "availability.json")
  out_csv <- if (length(args) >= 3) args[3] else file.path("data", "availability_long.csv")

  prepare_availability(in_xlsx, out_json, out_csv)
}

File: prepare_availability.py
#!/usr/bin/env python3
"""
Convert scraper output calendar_results.xlsx into:
  1) availability_long.csv  - flat/debug table
  2) availability.json      - compact structure for the Leaflet popup calendar

Usage:
  python prepare_availability.py calendar_results.xlsx availability.json availability_long.csv

No external Python packages required.
"""

from __future__ import annotations

import csv
import json
import re
import sys
import unicodedata
import zipfile
import xml.etree.ElementTree as ET
from collections import defaultdict
from datetime import datetime, date
from pathlib import Path
from typing import Any

NS = {"a": "http://schemas.openxmlformats.org/spreadsheetml/2006/main"}


def _excel_col_to_idx(cell_ref: str) -> tuple[int, int]:
    m = re.match(r"([A-Z]+)(\d+)", cell_ref)
    if not m:
        raise ValueError(f"Bad Excel cell reference: {cell_ref}")
    col = 0
    for ch in m.group(1):
        col = col * 26 + ord(ch) - 64
    return col - 1, int(m.group(2)) - 1


def _read_shared_strings(z: zipfile.ZipFile) -> list[str]:
    if "xl/sharedStrings.xml" not in z.namelist():
        return []
    root = ET.fromstring(z.read("xl/sharedStrings.xml"))
    out = []
    for si in root.findall("a:si", NS):
        texts = [t.text or "" for t in si.findall(".//a:t", NS)]
        out.append("".join(texts))
    return out


def _cell_value(cell: ET.Element, shared_strings: list[str]) -> Any:
    cell_type = cell.attrib.get("t")

    if cell_type == "inlineStr":
        return "".join(t.text or "" for t in cell.findall(".//a:t", NS))

    v = cell.find("a:v", NS)
    if v is None:
        return None

    raw = v.text
    if raw is None:
        return None

    if cell_type == "s":
        return shared_strings[int(raw)]
    if cell_type == "b":
        return bool(int(raw))
    if cell_type == "str":
        return raw

    # Numeric cell. Preserve integer-looking values as int.
    try:
        if "." in raw:
            num = float(raw)
            return int(num) if num.is_integer() else num
        return int(raw)
    except ValueError:
        return raw


def read_first_sheet_xlsx(path: str | Path) -> list[dict[str, Any]]:
    """Read the first worksheet into a list of dict rows."""
    with zipfile.ZipFile(path) as z:
        shared_strings = _read_shared_strings(z)
        # For this scraper output the first worksheet is sheet1.xml.
        sheet_xml = z.read("xl/worksheets/sheet1.xml")

    root = ET.fromstring(sheet_xml)

    rows: list[list[Any]] = []
    max_col = 0

    for row in root.findall(".//a:sheetData/a:row", NS):
        row_idx = int(row.attrib["r"]) - 1
        values_by_col: dict[int, Any] = {}

        for cell in row.findall("a:c", NS):
            col_idx, _ = _excel_col_to_idx(cell.attrib["r"])
            values_by_col[col_idx] = _cell_value(cell, shared_strings)
            max_col = max(max_col, col_idx)

        while len(rows) <= row_idx:
            rows.append([])
        rows[row_idx] = [values_by_col.get(i) for i in range(max_col + 1)]

    if not rows:
        return []

    headers = [str(x).strip() if x is not None else "" for x in rows[0]]
    dict_rows = []
    for row in rows[1:]:
        row = row + [None] * (len(headers) - len(row))
        dict_rows.append(dict(zip(headers, row)))
    return dict_rows


def slugify_hut_id(raw_hut: str) -> str:
    """Make a stable hut_id, e.g. 'Adolf-Noßberger-Hütte, AT' -> 'adolf-nossberger-hutte-at'."""
    x = raw_hut.strip()
    replacements = {
        "ß": "ss", "ẞ": "ss",
        "ä": "a", "ö": "o", "ü": "u",
        "Ä": "a", "Ö": "o", "Ü": "u",
    }
    for old, new in replacements.items():
        x = x.replace(old, new)
    x = unicodedata.normalize("NFKD", x).encode("ascii", "ignore").decode("ascii")
    x = x.lower()
    x = re.sub(r"[^a-z0-9]+", "-", x).strip("-")
    return x


def split_hut_country(raw_hut: str) -> tuple[str, str | None]:
    """Split 'Adamek-Hütte, AT' -> ('Adamek-Hütte', 'AT')."""
    m = re.match(r"^(.*?),\s*([A-Z]{2})$", raw_hut.strip())
    if m:
        return m.group(1), m.group(2)
    return raw_hut.strip(), None


def parse_calendar_date(month_header: Any, day: Any) -> str | None:
    """Convert month_header='7/2026', day=14 into '2026-07-14'."""
    if month_header in (None, "") or day in (None, ""):
        return None

    mh = str(month_header).strip()
    m = re.match(r"^(\d{1,2})/(\d{4})$", mh)
    if not m:
        raise ValueError(f"Unexpected month_header: {month_header!r}")

    month = int(m.group(1))
    year = int(m.group(2))
    return date(year, month, int(day)).isoformat()


def normalize_status(raw_status: Any, free_places: Any, disabled: Any) -> tuple[str, str, int | None]:
    """
    Return (status, level, free_places_clean).

    status = semantic state used in the popup
    level  = visual bucket used for colors
    """
    raw_status = str(raw_status or "").strip().lower()

    if free_places in (None, ""):
        free_clean = None
    else:
        free_clean = int(free_places)

    if raw_status == "error":
        return "error", "error", free_clean

    if disabled is True:
        return "closed", "closed", free_clean

    if raw_status == "unknown" or free_clean is None:
        return "unknown", "unknown", None

    if free_clean <= 0:
        return "full", "full", 0

    # Keep status semantically available; use level for color intensity.
    if free_clean <= 3:
        return "available", "low", free_clean
    if free_clean <= 9:
        return "available", "medium", free_clean
    return "available", "high", free_clean


def build_outputs(rows: list[dict[str, Any]]) -> tuple[list[dict[str, Any]], dict[str, Any]]:
    generated_at = datetime.now().replace(microsecond=0).isoformat()

    long_rows: list[dict[str, Any]] = []
    errors: dict[str, list[dict[str, Any]]] = defaultdict(list)
    huts: dict[str, dict[str, Any]] = {}

    for r in rows:
        raw_hut = r.get("Hut")
        if not raw_hut:
            continue

        raw_hut = str(raw_hut).strip()
        hut_id = slugify_hut_id(raw_hut)
        hut_name, country = split_hut_country(raw_hut)
        raw_status = r.get("status")

        # Scraper-level error rows do not represent a calendar day.
        if str(raw_status).lower() == "error":
            errors[hut_id].append({
                "hut": raw_hut,
                "message": r.get("error_message"),
            })
            continue

        iso_date = parse_calendar_date(r.get("month_header"), r.get("day"))
        if iso_date is None:
            continue

        status, level, free_clean = normalize_status(raw_status, r.get("free_places"), r.get("disabled"))

        huts.setdefault(hut_id, {
            "hut": raw_hut,
            "name": hut_name,
            "country": country,
            "calendar": {},
        })

        day_obj = {
            "free": free_clean,
            "status": status,
            "level": level,
        }

        # Keep debug fields only if useful for later troubleshooting.
        if r.get("raw_text") is not None:
            day_obj["raw_text"] = r.get("raw_text")
        if r.get("aria") is not None:
            day_obj["aria"] = r.get("aria")

        huts[hut_id]["calendar"][iso_date] = day_obj

        long_rows.append({
            "hut_id": hut_id,
            "hut_name": hut_name,
            "country": country,
            "date": iso_date,
            "free_places": free_clean,
            "status": status,
            "level": level,
            "raw_hut": raw_hut,
            "raw_status": raw_status,
            "raw_text": r.get("raw_text"),
            "aria": r.get("aria"),
            "disabled": r.get("disabled"),
            "color": r.get("color"),
        })

    all_dates = sorted({r["date"] for r in long_rows})

    # Add summaries for popup / marker coloring.
    for hut_id, h in huts.items():
        calendar = h["calendar"]
        available = [
            (d, v["free"])
            for d, v in calendar.items()
            if v.get("status") == "available" and isinstance(v.get("free"), int) and v.get("free") > 0
        ]
        available_sorted = sorted(available)
        h["summary"] = {
            "days_total": len(calendar),
            "days_available": len(available_sorted),
            "days_full": sum(1 for v in calendar.values() if v.get("status") == "full"),
            "days_unknown": sum(1 for v in calendar.values() if v.get("status") == "unknown"),
            "next_available_date": available_sorted[0][0] if available_sorted else None,
            "next_available_free": available_sorted[0][1] if available_sorted else None,
            "max_free_places": max((free for _, free in available_sorted), default=None),
            "total_free_place_days": sum(free for _, free in available_sorted),
        }

    availability_json = {
        "generated_at": generated_at,
        "date_from": all_dates[0] if all_dates else None,
        "date_to": all_dates[-1] if all_dates else None,
        "errors": dict(errors),
        "huts": dict(sorted(huts.items())),
    }

    return long_rows, availability_json


def write_long_csv(rows: list[dict[str, Any]], path: str | Path) -> None:
    fieldnames = [
        "hut_id", "hut_name", "country", "date", "free_places", "status", "level",
        "raw_hut", "raw_status", "raw_text", "aria", "disabled", "color",
    ]
    with open(path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)


def main() -> None:
    in_xlsx = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("calendar_results.xlsx")
    out_json = Path(sys.argv[2]) if len(sys.argv) > 2 else Path("availability.json")
    out_csv = Path(sys.argv[3]) if len(sys.argv) > 3 else Path("availability_long.csv")

    rows = read_first_sheet_xlsx(in_xlsx)
    long_rows, availability_json = build_outputs(rows)

    write_long_csv(long_rows, out_csv)
    with open(out_json, "w", encoding="utf-8") as f:
        json.dump(availability_json, f, ensure_ascii=False, indent=2, sort_keys=True)

    print(f"Input rows: {len(rows)}")
    print(f"Calendar rows written: {len(long_rows)}")
    print(f"Huts: {len(availability_json['huts'])}")
    print(f"Date range: {availability_json['date_from']} – {availability_json['date_to']}")
    print(f"Errors: {sum(len(v) for v in availability_json['errors'].values())}")
    print(f"Wrote: {out_json}")
    print(f"Wrote: {out_csv}")


if __name__ == "__main__":
    main()
